nexus_constructor/component_tree_model.py

```diff
@@ -0,0 +1,284 @@
+    def __init__(self, parent):
+        super().__init__()
+        self.parent = parent
+
+
+class LinkTransformation:
+    def __init__(self, parent):
+        super().__init__()
+        self.parent = parent
+        self.link_transformation = None
+
+
+class ComponentTreeModel(QAbstractItemModel):
+    def __init__(self, instrument, parent=None):
+        super().__init__(parent)
+        self.instrument = instrument
+        self.components = self.instrument.get_component_list()
+
+    def columnCount(self, parent):
+        return 1
+
+    def data(self, index, role):
+        if not index.isValid():
+            return None
+        item = index.internalPointer()
+        if role == Qt.DisplayRole:
+            return item
+        elif role == Qt.SizeHintRole:
+            return
+
+    def flags(self, index):
+        if not index.isValid():
+            return Qt.NoItemFlags
+        parentItem = index.internalPointer()
+        if issubclass(type(parentItem), ComponentModel):
+            return Qt.ItemIsEnabled | Qt.ItemIsSelectable
+        elif issubclass(type(parentItem), ComponentInfo):
+            return Qt.ItemIsEnabled
+        elif type(parentItem) is TransformationsList:
+            return Qt.ItemIsEnabled | Qt.ItemIsSelectable
+        return Qt.ItemIsEnabled | Qt.ItemIsSelectable | Qt.ItemIsEditable
+
+    def supportedDropActions(self) -> PySide2.QtCore.Qt.DropActions:
+        return Qt.DropAction.MoveAction
+
+    def add_link(self, node):
+        parentItem = node.internalPointer()
+        transformation_list = None
+        target_index = QModelIndex()
+        if isinstance(parentItem, ComponentModel):
+            if not hasattr(parentItem, "stored_transforms"):
+                parentItem.stored_transforms = parentItem.transforms
+            transformation_list = parentItem.stored_transforms
+            target_index = self.index(1, 0, node)
+        elif isinstance(parentItem, TransformationsList):
+            transformation_list = parentItem
+            target_index = node
+        elif isinstance(parentItem, TransformationModel):
+            transformation_list = parentItem.parent
+            target_index = self.parent(node)
+        if transformation_list.has_link:
+            return
+        target_pos = len(transformation_list)
+        self.beginInsertRows(target_index, target_pos, target_pos)
+        transformation_link = LinkTransformation(transformation_list)
+        transformation_list.has_link = True
+        transformation_list.link = transformation_link
+        self.endInsertRows()
+
+    def add_component(self, new_component):
+        self.beginInsertRows(QModelIndex(), len(self.components), len(self.components))
+        self.components.append(new_component)
+        self.endInsertRows()
+
+    def remove_node(self, node):
+        if isinstance(node.internalPointer(), ComponentModel):
+            remove_index = self.components.index(node.internalPointer())
```

Comment:
I suggest refactoring out `self.__remove_component()`, `self.__remove_transformation()` and `self.__remove_link()`.

Reply:
Done.